sdks/python/apache_beam/transforms/stats.py

```diff
@@ -0,0 +1,206 @@
+
+
+class _LargestUnique(object):
+  """
+  An object to keep samples and calculate sample hash space. It is an
+  accumulator of a combine function.
+  """
+  _HASH_SPACE_SIZE = 2.0 * sys.maxsize
+
+  def __init__(self, sample_size):
+    self._sample_size = sample_size
+    self._min_hash = sys.maxsize
+    self._sample_heap = []
+    self._sample_set = set()
+
+  def add(self, element):
+    """
+    :param an element from pcoll.
+    :return: boolean type whether the value is in the heap
+
+    Adds a value to the heap, returning whether the value is (large enough to
+    be) in the heap.
+    """
+    if len(self._sample_heap) >= self._sample_size and element < self._min_hash:
+      return False
+
+    if element not in self._sample_set:
+      self._sample_set.add(element)
+      heapq.heappush(self._sample_heap, element)
+
+      if len(self._sample_heap) > self._sample_size:
+        temp = heapq.heappop(self._sample_heap)
+        self._sample_set.remove(temp)
+        self._min_hash = self._sample_heap[0]
+      elif element < self._min_hash:
+        self._min_hash = element
+
+    return True
+
+  def get_estimate(self):
+    """
+    :return: estimation of unique values
+
+    If heap size is smaller than sample size, just return heap size.
+    Otherwise, takes into account the possibility of hash collisions,
+    which become more likely than not for 2^32 distinct elements.
+    Note that log(1+x) ~ x for small x, so for sampleSize << maxHash
+    log(1 - sampleSize/sampleSpace) / log(1 - 1/sampleSpace) ~ sampleSize
+    and hence estimate ~ sampleSize * HASH_SPACE_SIZE / sampleSpace
+    as one would expect.
+    """
+
+    if len(self._sample_heap) < self._sample_size:
+      return len(self._sample_heap)
+    else:
+      sample_space_size = sys.maxsize - 1.0 * self._min_hash
+      est = math.log1p(-self._sample_size/ sample_space_size) \
+            / math.log1p(-1 / sample_space_size) \
+            * self._HASH_SPACE_SIZE \
+            / sample_space_size
+
+      return round(est)
+
+
+class ApproximateUniqueCombineDoFn(CombineFn):
+  """
+  ApproximateUniqueCombineDoFn computes an estimate of the number of
+  unique values that were combined.
+  """
+
+  def __init__(self, sample_size):
+    self._sample_size = sample_size
+
+  def create_accumulator(self, *args, **kwargs):
+    return _LargestUnique(self._sample_size)
+
+  @staticmethod
+  def add_input(accumulator, element, *args, **kwargs):
+    try:
+      if 'mmh3' in sys.modules:
+        accumulator.add(mmh3.hash64(str(element))[1])
+      else:
+        accumulator.add(hash(str(element)))
+      return accumulator
+    except Exception as e:
+      raise RuntimeError("Runtime exception: %s", e)
+
+  def merge_accumulators(self, accumulators, *args, **kwargs):
+    merged_accumulator = self.create_accumulator()
+    for accumulator in accumulators:
+      for i in accumulator._sample_heap:
+        merged_accumulator.add(i)
```

Comment:
Is it possible to offer a fast merge for 2 accumulators? Like a merge two heaps in a single call? Perhaps file a JIRA for future optimization of this.